File: 01-programming-languages/C#/01-fundamentals/05-arrays/arrays.cs
// C# Arrays - Array, List, Dictionary

using System;
using System.Collections.Generic;

namespace ArraysDemo
{
    class Program
    {
        static void Main(string[] args)
        {
            // Arrays - fixed size
            Console.WriteLine("=== Arrays ===");
            int[] numbers = new int[5];
            numbers[0] = 1;
            numbers[1] = 2;
            numbers[2] = 3;
            numbers[3] = 4;
            numbers[4] = 5;
            
            string[] fruits = { "Apple", "Banana", "Cherry" };
            
            // Array methods
            Console.WriteLine($"Length: {fruits.Length}");
            Console.WriteLine($"First: {fruits[0]}");
            Console.WriteLine($"Last: {fruits[^1]}"); // Index from end
            
            // Array traversal
            Console.Write("Fruits: ");
            foreach (string fruit in fruits)
            {
                Console.Write($"{fruit} ");
            }
            Console.WriteLine();
            
            // Array.Sort
            Array.Sort(numbers);
            
            // Array.IndexOf
            int index = Array.IndexOf(fruits, "Banana");
            Console.WriteLine($"Banana index: {index}");
            
            // List<T> - dynamic size
            Console.WriteLine("\n=== List ===");
            List<string> names = new List<string>();
            names.Add("Alice");
            names.Add("Bob");
            names.Add("Charlie");
            
            // List methods
            Console.WriteLine($"Count: {names.Count}");
            Console.WriteLine($"Contains Bob: {names.Contains("Bob")}");
            
            names.Insert(1, "Bobby");
            names.Remove("Charlie");
            
            Console.Write("Names: ");
            foreach (string name in names)
            {
                Console.Write($"{name} ");
            }
            Console.WriteLine();
            
            // List with initial capacity
            List<int> largeList = new List<int>(1000);
            
            // List.Find
            string found = names.Find(name => name.StartsWith("B"));
            Console.WriteLine($"Found: {found}");
            
            // List.RemoveAll
            names.RemoveAll(name => name.Length > 5);
            
            // Dictionary<TKey, TValue> - key-value pairs
            Console.WriteLine("\n=== Dictionary ===");
            Dictionary<string, int> ages = new Dictionary<string, int>();
            ages["Alice"] = 30;
            ages["Bob"] = 25;
            ages["Charlie"] = 35;
            
            // Dictionary methods
            Console.WriteLine($"Alice's age: {ages["Alice"]}");
            Console.WriteLine($"Contains Bob: {ages.ContainsKey("Bob")}");
            Console.WriteLine($"Contains age 30: {ages.ContainsValue(30)}");
            
            // TryGetValue
            if (ages.TryGetValue("David", out int age))
            {
                Console.WriteLine($"David's age: {age}");
            }
            else
            {
                Console.WriteLine("David not found");
            }
            
            // Dictionary traversal
            Console.WriteLine("Ages:");
            foreach (KeyValuePair<string, int> kvp in ages)
            {
                Console.WriteLine($"  {kvp.Key}: {kvp.Value}");
            }
            
            // Remove from dictionary
            ages.Remove("Charlie");
            
            // Tuple
            Console.WriteLine("\n=== Tuples ===");
            var person = (Name: "Alice", Age: 30);
            Console.WriteLine($"{person.Name}: {person.Age}");
            
            // Deconstruction
            var (name, age) = person;
            Console.WriteLine($"Deconstructed: {name}, {age}");
            
            Console.WriteLine("Arrays example running");
        }
    }
}

File: 01-programming-languages/JavaScript/01-fundamentals/05-arrays/arrays.js
/*
 * JavaScript Arrays
 * Array methods: map, filter, reduce, forEach
 */

// ============================================
// Creating Arrays
// ============================================

const fruits = ["apple", "banana", "cherry"];
const numbers = [1, 2, 3, 4, 5];
const mixed = [1, "two", true, null, { key: "value" }];
const empty = [];

// Array constructor
const arr1 = new Array(5); // Creates array with 5 empty slots
const arr2 = Array.from("Hello"); // ["H", "e", "l", "l", "o"]
const arr3 = Array.of(1, 2, 3); // [1, 2, 3]

// ============================================
// Accessing and Modifying
// ============================================

console.log(fruits[0]);      // "apple"
console.log(fruits.length);  // 3
fruits[1] = "blueberry";     // Modify element
fruits.push("date");         // Add to end
fruits.unshift("avocado");   // Add to start
const last = fruits.pop();   // Remove from end
const first = fruits.shift(); // Remove from start

// ============================================
// map() - Transform each element
// ============================================

const doubled = numbers.map(num => num * 2);
console.log("Doubled:", doubled); // [2, 4, 6, 8, 10]

const names = fruits.map(fruit => fruit.toUpperCase());
console.log("Uppercase:", names);

// ============================================
// filter() - Select elements matching condition
// ============================================

const evens = numbers.filter(num => num % 2 === 0);
console.log("Evens:", evens); // [2, 4]

const longFruits = fruits.filter(fruit => fruit.length > 5);
console.log("Long fruits:", longFruits);

// ============================================
// reduce() - Accumulate into single value
// ============================================

const sum = numbers.reduce((accumulator, current) => {
    return accumulator + current;
}, 0);
console.log("Sum:", sum); // 15

const wordLengths = fruits.reduce((acc, fruit) => {
    acc[fruit] = fruit.length;
    return acc;
}, {});
console.log("Word lengths:", wordLengths);

// ============================================
// forEach() - Execute function for each element
// ============================================

numbers.forEach((num, index) => {
    console.log(`Index ${index}: ${num}`);
});

// ============================================
// find() and findIndex()
// ============================================

const found = numbers.find(num => num > 3);
console.log("Found:", found); // 4

const foundIndex = numbers.findIndex(num => num > 3);
console.log("Found index:", foundIndex); // 3

// ============================================
// some() and every()
// ============================================

const hasNegative = numbers.some(num => num < 0);
console.log("Has negative:", hasNegative); // false

const allPositive = numbers.every(num => num > 0);
console.log("All positive:", allPositive); // true

// ============================================
// flat() and flatMap()
// ============================================

const nested = [1, [2, 3], [4, [5, 6]]];
const flat = nested.flat(Infinity);
console.log("Flat:", flat); // [1, 2, 3, 4, 5, 6]

const sentences = ["Hello World", "Goodbye Moon"];
const words = sentences.flatMap(sentence => sentence.split(" "));
console.log("Words:", words);

// ============================================
// includes(), indexOf(), slice(), splice()
// ============================================

console.log("Includes 3:", numbers.includes(3)); // true
console.log("Index of 3:", numbers.indexOf(3)); // 2

const sliced = numbers.slice(1, 4); // [2, 3, 4]
console.log("Sliced:", sliced);

const removed = numbers.splice(2, 1); // Remove 1 element at index 2
console.log("Removed:", removed); // [3]
console.log("After splice:", numbers);

// ============================================
// sort()
// ============================================

const unsorted = [3, 1, 4, 1, 5, 9, 2, 6];
const sorted = [...unsorted].sort((a, b) => a - b); // Ascending
console.log("Sorted:", sorted);

const sortedDesc = [...unsorted].sort((a, b) => b - a); // Descending
console.log("Sorted desc:", sortedDesc);
